src/neuralnet/researcher.cpp

```diff
@@ -0,0 +1,483 @@
+extern std::string msMiningErrors;
+
+namespace {
+//!
+//! \brief Global BOINC researcher mining context.
+//!
+ResearcherPtr researcher = std::make_shared<Researcher>();
+
+//!
+//! \brief Convert a project name to lowercase change any underscores to spaces.
+//!
+//! NOTICE: Because the application merges the loaded BOINC project details into
+//! the output of the "projects" RPC method, project whitelist contracts created
+//! by the master key holder must embed the project names into the contract just
+//! as output by this function or the method will omit projects with mismatching
+//! names from the result.
+//!
+//! \return A "normalized" project name.
+//!
+std::string LowerUnderscore(std::string data)
+{
+    boost::to_lower(data);
+    boost::replace_all(data, "_", " ");
+
+    return data;
+}
+
+//!
+//! \brief Determine whether the wallet must run in investor mode before trying
+//! to load BOINC CPIDs.
+//!
+//! \return \c true if the user explicitly configured investor mode or failed
+//! to input a valid email address.
+//!
+bool ConfiguredForInvestorMode()
+{
+    if (GetBoolArg("investor", false)) {
+        LogPrintf("Investor mode configured. Skipping CPID import.");
+        return true;
+    }
+
+    if (Researcher::Email().empty()) {
+        LogPrintf(
+            "WARNING: Please set 'email=<your BOINC account email>' in "
+            "gridcoinresearch.conf. Continuing in investor mode.");
+
+        return true;
+    }
+
+    return false;
+}
+
+//!
+//! \brief Fetch the contents of BOINC's client_state.xml file from disk.
+//!
+//! \return The entire client_state.xml file as a string if readable.
+//!
+boost::optional<std::string> ReadClientStateXml()
+{
+    const std::string path = GetBoincDataDir();
+    const std::string contents = GetFileContents(path + "client_state.xml");
+
+    if (contents != "-1") {
+        return contents;
+    }
+
+    LogPrintf("WARNING: Unable to obtain BOINC CPIDs.");
+
+    if (!GetArgument("boincdatadir", "").empty()) {
+        LogPrintf("Could not access configured BOINC data directory %s", path);
+    } else {
+        LogPrintf(
+            "BOINC data directory is not installed in the default location.\n"
+            "Please specify its current location in gridcoinresearch.conf.");
+    }
+
+    return boost::none;
+}
+
+//!
+//! \brief Load a set of project XML sections from BOINC's client_state.xml
+//! file to gather CPIDs from.
+//!
+//! \return Items containing the XML between each \c <project>...</project>
+//! section present in the client_state.xml file or an empty container when
+//! the file is not accessible or contains no project elements.
+//!
+std::vector<std::string> FetchProjectsXml()
+{
+    const boost::optional<std::string> client_state_xml = ReadClientStateXml();
+
+    if (!client_state_xml) {
+        return { };
+    }
+
+    std::vector<std::string> projects = split(*client_state_xml, "<project>");
+
+    if (projects.size() < 2) {
```

Comment:
What happens if you only have one project?

Reply:
The first element from the split is just a bunch of crap from the beginning of the file. It doesn't contain a project, so we drop it right below this.

Reply:
If `split()` returns one item, it's the whole file and doesn't contain any projects.